File: Java/2337. Move Pieces to Obtain a String.java
class Solution {
    public boolean canChange(String start, String target) {
        int p1 = 0, p2 = 0, size = start.length();

        while (p1 < size || p2 < size) {
            while (p1 < size && start.charAt(p1) == '_') {
                p1++;
            }

            while (p2 < size && target.charAt(p2) == '_') {
                p2++;
            }

            if (p1 == size || p2 == size) {
                return p1 == size && p2 == size;
            }

            if (start.charAt(p1) != target.charAt(p2)) {
                return false;
            }

            if (p1 > p2 && start.charAt(p1) == 'R' || p1 < p2 && start.charAt(p1) == 'L') {
                return false;
            }

            p1++;
            p2++;
        }
        return true;
    }
}

File: 2337. Move Pieces to Obtain a String.py
class Solution:
    def canChange(self, start: str, target: str) -> bool:
        p1 = p2 = 0
        size = len(start)

        while p1 < size or p2 < size:
            # Situations where one of them is space
            while p1 < size and start[p1] == '_':
                p1 += 1

            while p2 < size and target[p2] == '_':
                p2 += 1

            # Both pointers should end at the same time
            if p1 == size or p2 == size:
                return p1 == size and p2 == size

            # Skipped all the spaces, the two chars should at least be the same
            if start[p1] != target[p2]:
                return False

            # Although the chars are the same, they might not be able to get moved correctly
            if p1 > p2 and start[p1] == 'R':
                return False
            
            if p1 < p2 and start[p1] == 'L':
                return False

            p1 += 1
            p2 += 1

        return True